install: support config files

bgilbert opened the pull request:
Add a repeatable `--config-file` option which specifies a YAML config file.  This contains command-line arguments that are prepended to the actual `coreos-installer install` command line before execution.  This will allow users to configure the install process without resorting to kernel arguments or replacing the `coreos-installer.service` unit.

Make the target device argument optional if a config file is provided, since the config file might specify it instead, but also do a late check in case the config file doesn't provide one either.

Implement config merging by generating a vector of synthesized command-line arguments, prepending it to the actual command line, and having structopt reparse the whole thing.  That allows structopt to apply any validation and constraints specified in struct attributes.

jlebon commented:
I wonder if a simpler alternative would be to first check for `--config-file` args, and if found, rerun `Cmd::from_iter()`, pointing it to a `Vec` of what we read in + everything else that's left over? (And do this iteratively). That would avoid all the serialization goop I think.

jlebon commented on src/cmdline.rs:
```diff
@@ -586,10 +686,428 @@ impl DefaultString for Architecture {
+
+    /// Test multiple config files overlapping with command-line arguments
+    #[test]
+    fn install_config_file_overlapping_field() {
```
Would be good to expand this test (or add a separate test) with a YAML doc containing all the fields. Can reuse the values from `serialize_full_install_config` so we check the deserialized version against the `InstallConfig` there.

jlebon commented on docs/cmd/install.md:
```diff
@@ -19,6 +19,7 @@ Install Fedora CoreOS or RHEL CoreOS
 
 ## Options
 
+| **--config-file**, **-c** *path* | YAML config file with install options |
```
Will need more docs for this describing the format.

jlebon commented on src/cmdline.rs:
```diff
@@ -586,10 +686,428 @@ impl DefaultString for Architecture {
+        }
+    }
+
+    // enormous pile of boilerplate
```
Was most of this copy/pasted from some serde doc? Should we link it here?

jlebon commented on src/cmdline.rs:
```diff
@@ -215,21 +233,80 @@ pub struct InstallConfig {
     pub fetch_retries: FetchRetries,
 
     // positional args
+    // use #[serde(rename(serialize = "_[...]"))] on each of these
```
Can you expand that comment with the why? Basically something like:

```suggestion
    // use #[serde(rename(serialize = "_[...]"))] on each of these so that
    // we don't serialize it as e.g. `--device DEV`
```

jlebon reviewed (approved):
Some minor comments but LGTM overall. The amount of boilerplate involved isn't great, and there's a maintenance cost there, though I don't have a better idea if we want to retain the nicer YAML UX (instead of just a pure list of CLI args).

bgilbert commented:
Discussed this some OOB with @jlebon.  We could dramatically reduce the serde attribute goo/boilerplate if the config file was just a list of command-line arguments, without schema enforcement, but I hesitate to do that.  We'll be encouraging users to write and maintain these config files, so it's nice to have a format that's a little more structured.  I've tried to implement this PR in a way that minimizes ongoing maintenance burden (e.g. by reusing field stringification/destringification needed by structopt), and I've also now added a bunch of comments for context, but there's no doubt that the current approach is a more cumbersome implementation.

I did also try directly merging `InstallConfig` structs, rather than serializing them back to command-line arguments.  That turned out to be worse, especially since structopt can't handle the constraint enforcement in that model.  A lot of this would be easier if we could use a proc macro, but those currently can't be implemented and invoked within the same crate, and it doesn't seem worth adding a second crate for this.

Updated for feedback.  Additional changes in this version:

- Added user-facing config file docs.
- Config files can now override each other, and command-line arguments can override config files.
- A later PR will need clean serialization from `InstallConfig` back to YAML, requiring more struct attributes and a worse way of detecting positional parameters (boo).
- Modified `coreos-installer-service` to run if there are files in `/etc/coreos/installer.d`, and to pass those files to coreos-installer as config files.
- Renamed the `device` config file setting to `dest-device` for clarity.

jlebon reviewed (approved):
Nice work! This looks good overall, just some minor items.

jlebon commented:
We should also add some coverage for this in `kola testiso`.

bgilbert commented on scripts/coreos-installer-service:
```diff
@@ -44,17 +47,29 @@ copy_arg() {
 
+# Config files
+have_config_file=
+for file in ${INSTALLER_CONFIG_DIR}/*; do
```
I don't see any benefit to requiring a particular extension, and it adds a footgun.  We don't need to worry about RPM backup files, editor backup files, etc., because the files will be placed in the directory by Ignition, exist for one boot, and then be deleted.  The footgun is that files lacking the extension would be ignored, which could be unpleasant for users to debug.

jlebon replied:
Don't feel strongly about this, but IMO the benefit is simply that it's more descriptive and enforces consistency.

> The footgun is that files lacking the extension would be ignored, which could be unpleasant for users to debug.

Hmm, I'm not sure how much of a concern that is. systemd does the same for many (all?) of its dropins directories like tmpfiles.d, sysusers.d, etc...

bgilbert replied:
I looked at a bunch of packages with `.d` directories in `/etc` on my system, and results were mixed:

### Require extension

- auditd
- dbus
- dhclient
- dnf
- dracut
- fwupd
- kmod
- ld.so (RPM package)
- NetworkManager
- ostree
- polkit
- sshd (RPM package)
- Xorg

### Do not require extension

- bash
- dconf
- dnsmasq (RPM package)
- grub
- kdump
- logrotate
- PAM
- sudo

It looks like there may be a trend in newer tools of requiring an extension, which makes sense for RPM packages that ship conffiles.  In this case I'm still not convinced that the extra debug overhead (for admins who may not be familiar with the convention) is worth it, though.  (And it's inconsistent with the `--config-file` option, which doesn't care about extensions.)

I'm going to go ahead and merge for now, and we can revisit in a followup if desired.

bgilbert commented:
Yeah, actually, this will let us clean up some hacks in `kola testiso`.  I'll add that to the queue.

Updated the docs page.
